source/lac/trilinos_epetra_vector.cc

```diff
@@ -0,0 +1,529 @@
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LinearAlgebra
+{
+  namespace EpetraWrappers
+  {
+    Vector::Vector()
+      :
+      vector(new Epetra_FEVector(Epetra_Map(0,1,0,Epetra_MpiComm(MPI_COMM_SELF))))
+    {}
+
+
+
+    Vector::Vector(const Vector &V)
+      :
+      vector(new Epetra_FEVector(V.trilinos_vector()))
+    {}
+
+
+
+    Vector::Vector(const IndexSet &parallel_partitioner,
+                   const MPI_Comm &communicator)
+      :
+      vector(new Epetra_FEVector(parallel_partitioner.make_trilinos_map(communicator,false)))
+    {}
+
+
+
+    void Vector::reinit(const IndexSet &parallel_partitioner,
+                        const MPI_Comm &communicator,
+                        const bool      omit_zeroing_entries)
+    {
+      Epetra_Map input_map = parallel_partitioner.make_trilinos_map(communicator,false);
+      if (vector->Map().SameAs(input_map)==false)
+        vector.reset(new Epetra_FEVector(input_map));
+      else if (omit_zeroing_entries==false)
+        {
+          const int ierr = vector->PutScalar(0.);
+          (void) ierr;
+          Assert(ierr==0, ExcTrilinosError(ierr));
+        }
+    }
+
+
+
+    Vector &Vector::operator= (const Vector &V)
+    {
+      // Distinguish three cases:
+      //  - First case: both vectors have the same layout.
+      //  - Second case: both vectors have the same size but different layout.
+      //  - Third case: the vectors have different size.
+      if (vector->Map().SameAs(V.trilinos_vector().Map()))
+        *vector = V.trilinos_vector();
+      else
+        {
+          if (size()==V.size())
+            {
+              Epetra_Import data_exchange(vector->Map(), V.trilinos_vector().Map());
+
+              const int ierr = vector->Import(V.trilinos_vector(), data_exchange, Insert);
+              Assert(ierr==0, ExcTrilinosError(ierr));
+            }
+          else
+            vector.reset(new Epetra_FEVector(V.trilinos_vector()));
+        }
+
+      return *this;
+    }
+
+
+
+    Vector &Vector::operator= (const ::dealii::LinearAlgebra::ReadWriteVector<double> &V)
+    {
+      IndexSet elements_to_write(V.get_stored_elements());
+      IndexSet locally_owned_elements(this->locally_owned_elements());
+      IndexSet off_proc_elements(elements_to_write);
+      off_proc_elements.subtract_set(locally_owned_elements);
+      IndexSet on_proc_elements(elements_to_write);
+      on_proc_elements.subtract_set(off_proc_elements);
+
+      // Write the elements that locally owned.
+      IndexSet::ElementIterator elem = on_proc_elements.begin(),
+                                elem_end = on_proc_elements.end();
+      for (; elem!=elem_end; ++elem)
+        {
+          const TrilinosWrappers::types::int_type local_index =
+            vector->Map().LID(static_cast<TrilinosWrappers::types::int_type>(*elem));
+          (*vector)[0][local_index] = V[*elem];
+        }
+
+
+      // Write the elements off-processor. Cannot use Import function of Trilinos
```

Comment:
I have a few questions:
1. Why isn't it possible to construct an Epetra_Map from the index set of the source vector, initialize an Epetra_Vector with that map, set all the elements in the vector, and do Epetra_Import/Export (I always confuse which one to use in this context) to the *this vector? That sounds like a much simpler version and is likely to scale better.

2. I do not like to write algorithms that figure out the owner of ghost indices just from the vector entries in the general case. It was difficult enough to do for parallel::distributed::Vector (or MPI::Partitioner) in the simple case of contiguous locally owned ranges. I'd rather let external libraries do that work. I looked into the Epetra_MpiDistributor a year or two ago (that gets used in the general case IIRC) and decided that I do not want to understand how it manages to do the general data exchange.

3. You say in the in-code comment that this code isn't going to scale to 10000 processors. If I understand the code correctly I guess it will prominently show up already on 1000 processors. Assume you have 30,000 elements per processor in the vector and 5,000 ghosts. You communicate the ghost indices all-to-all, which means that with 1000 processors you already end up with 50m ghost entries. Exchanging 2x400m data per MPI rank just for one single vector is borderline already.